File: sliding_window/longest_substring_without_repeating_characters.py
import sys

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        indices = {}
        length = l = r = 0

        while r < len(s):
            # first check if the current character is in the dictionary
            # if found, we have a duplicate situation
            # if the saved index of the current character is greater than
            # the current position of the left pointer, we set the left
            # pointer to the saved index plus one.

            # if the left pointer's current position is greater than the saved
            # index of the current character, it means that we have previously
            # encountered a duplicate situation, IF we now move the left
            # pointer backwards, we will be including a duplicate section and
            # our substring will no longer be valid.

            if s[r] in indices and indices[s[r]] + 1 > l:
                l = indices[s[r]] + 1

            indices[s[r]] = r

            length = max(length, r - l + 1)

            r += 1

        return length

if __name__ == "__main__":
    s = Solution()
    print(s.lengthOfLongestSubstring(sys.argv[1]))

File: sliding_window/longest_substring_without_repeating_characters.cpp
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    int substring_start = 0, longest_length = 0;
    std::unordered_map<char, int> char_position;
    for (int i = 0; i < s.length(); i++) {
      if (char_position.count(s[i]) > 0) {
        substring_start = char_position[s[i]] + 1 > substring_start
                              ? char_position[s[i]] + 1
                              : substring_start;
      }
      char_position[s[i]] = i;
      longest_length = std::max(longest_length, i - substring_start + 1);
    }
    return longest_length;
  }
};
